Add initial integration for GridFormattingContext

yulunwu2 opened the pull request:
#### 1037e592cea83a9e83f0a6c5c5e61700a098dd82
<pre>
Add initial integration for GridFormattingContext
<a href="https://bugs.webkit.org/show_bug.cgi?id=298884">https://bugs.webkit.org/show_bug.cgi?id=298884</a>
&lt;<a href="https://rdar.apple.com/160627422">rdar://160627422</a>&gt;

Reviewed by Sammy Gill.

This PR adds the foundational integration layer for GridFormattingContext.

Specifically, this PR:

1. Adds the integration point for the new Grid Formatting Context in the form of
LayoutIntegrationGridLayout.
2. Introduces layoutUsingGridFormattingContext() and canUseForGridLayout() to gate
usage of the new layout code path.
3. Adds buildTreeForGridContent() to the the BoxTreeUpdater to build a tree of Layout::ElementBox.
4. Caches the top-level GridBox and LayoutState in GridFormattingContext.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/layout/formattingContexts/grid/GridFormattingContext.cpp:
(WebCore::Layout::GridFormattingContext::GridFormattingContext):
* Source/WebCore/layout/formattingContexts/grid/GridFormattingContext.h:
* Source/WebCore/layout/integration/LayoutIntegrationBoxTreeUpdater.cpp:
(WebCore::LayoutIntegration::BoxTreeUpdater::build):
(WebCore::LayoutIntegration::BoxTreeUpdater::buildTreeForGridContent):
* Source/WebCore/layout/integration/LayoutIntegrationBoxTreeUpdater.h:
* Source/WebCore/layout/integration/LayoutIntegrationCoverage.cpp:
(WebCore::LayoutIntegration::canUseForGridLayout):
* Source/WebCore/layout/integration/LayoutIntegrationCoverage.h:
* Source/WebCore/layout/integration/grid/LayoutIntegrationGridLayout.cpp: Copied from Source/WebCore/layout/formattingContexts/grid/GridFormattingContext.cpp.
(WebCore::LayoutIntegration::GridLayout::GridLayout):
(WebCore::LayoutIntegration::m_layoutState):
(WebCore::LayoutIntegration::GridLayout::layout):
* Source/WebCore/layout/integration/grid/LayoutIntegrationGridLayout.h: Copied from Source/WebCore/layout/formattingContexts/grid/GridFormattingContext.h.
(WebCore::LayoutIntegration::GridLayout::gridBox const):
(WebCore::LayoutIntegration::GridLayout::gridBox):
(WebCore::LayoutIntegration::GridLayout::gridBoxRenderer const):
(WebCore::LayoutIntegration::GridLayout::gridBoxRenderer):
(WebCore::LayoutIntegration::GridLayout::layoutState):
(WebCore::LayoutIntegration::GridLayout::layoutState const):
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::layoutBlock):
(WebCore::RenderGrid::layoutUsingGridFormattingContext):
* Source/WebCore/rendering/RenderGrid.h:

Canonical link: <a href="https://commits.webkit.org/300088@main">https://commits.webkit.org/300088@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/940e93dda73c335323a689942b32f7884f5da596

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121268 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/40965 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31623 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127704 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73349 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e2beafdc-1571-42f5-8243-5cdfbddd6ee0) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/123144 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41667 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49544 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92113 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61283 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/45ca0799-3562-4553-a65f-2a73286ce633) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124220 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33264 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108669 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72790 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d2cd4018-521a-4014-99ae-befad9aba7a4) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32280 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26797 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71286 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102757 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26974 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130542 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48196 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36632 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100711 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48563 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104859 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100616 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25508 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46022 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24082 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/44889 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48054 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53767 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47525 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50872 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49208 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebCore/layout/formattingContexts/grid/GridFormattingContext.h:
```diff
@@ -44,9 +44,13 @@ class GridFormattingContext : public CanMakeCheckedPtr<GridFormattingContext> {
+
+private:
+    const CheckedRef<const ElementBox> m_gridBox;
+    const CheckedRef<const LayoutState> m_globalLayoutState;
```
Should LayoutState actually be const? Unless we expect to only read from it we may end up mutating it during layout.

yulunwu2 replied:
I was thinking we could make it const for now and update this once we need to mutate it.

yulunwu2 replied:
Edit: making the fix.

sammygill commented on Source/WebCore/layout/formattingContexts/grid/GridFormattingContext.h:
```diff
@@ -25,9 +25,9 @@
 
 #pragma once
 
+#include "LayoutState.h"
```
Can we get away with just forward declaring this?

yulunwu2 replied:
Fixed!

sammygill commented on Source/WebCore/layout/integration/grid/LayoutIntegrationGridLayout.h:
```diff
@@ -0,0 +1,62 @@
+
+#pragma once
+
+#include "layout/LayoutState.h"
```
Should just be `LayoutState.h`

yulunwu2 replied:
Fixed!

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -370,6 +372,9 @@ void RenderGrid::computeLayoutRequirementsForItemsBeforeLayout(GridLayoutState&
 void RenderGrid::layoutBlock(RelayoutChildren relayoutChildren, LayoutUnit)
 {
+    if (layoutUsingGridFormattingContext())
+        return;
```
I think it might be better for this to live around/inside `layoutGrid()` for now so that we don't have to worry about the masonry case. If we want to or need to take it into consideration at some point we can but otherwise moving it there should hopefully make things more simple

yulunwu2 replied:
Fixed.

sammygill commented on Source/WebCore/rendering/RenderGrid.h:
```diff
@@ -302,6 +306,10 @@ class RenderGrid final : public RenderBlock {
+
+    bool layoutUsingGridFormattingContext();
+
+    std::optional<bool> m_hasGridFormattingContextLayout;
```
Do we need this currently or can we just get away with using `canUseForGridLayout` for now? My only concern is that if we add a member like this on the renderer I presume we would also need to think about when and how this gets invalidated in different scenarios.

yulunwu2 replied:
This isn't strictly necessary, but I don't think we need to worry about invalidating this.

From what I can tell from reading RenderFlexibleBox, m_hasFlexFormattingContextLayout is used to minimize the cost from repeatedly checking unsupported content and doesn't run into issues with m_hasXXXFormattingContextLayout being invalidated.

weinig commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -22376,6 +22376,20 @@
 		FFEFAB2D183BCC6F00514567 /* InternalObserverMap.cpp */ = {isa = PBXFileReference; lastKnownFileType = text; path = InternalObserverMap.cpp; sourceTree = "<group>"; };
 /* End PBXFileReference section */
 
+/* Begin PBXFileSystemSynchronizedBuildFileExceptionSet section */
```
What's going on with these Xcode project file changes? They look pretty odd for just adding two files.

sammygill commented on Source/WebCore/layout/integration/grid/LayoutIntegrationGridLayout.cpp:
```diff
@@ -0,0 +1,59 @@
+RenderGrid& GridLayout::gridBoxRenderer()
+{
+    return downcast<RenderGrid>(*m_gridBox->rendererForIntegration());
+}
```
I think you need to satisfy the static analyzer by either protecting the pointer here or making the member pointer const

https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines#mark-data-members-that-are-smart-pointers-as-const-if-they-never-get-reassigned

sammygill reviewed (commented):
Looks good overall I think we just need to solve the safer CPP issues and make the builds happy (might need to just add the mentioned files into `WebCore/Headers.cmake`

sammygill commented on Source/WebCore/layout/integration/grid/LayoutIntegrationGridLayout.h:
```diff
@@ -0,0 +1,63 @@
+    const RenderGrid& gridBoxRenderer() const { return downcast<RenderGrid>(*m_gridBox->rendererForIntegration()); }
+    RenderGrid& gridBoxRenderer() { return downcast<RenderGrid>(*m_gridBox->rendererForIntegration()); }
+
+
```
Extra new line?

yulunwu2 replied:
Yup, fixed

sammygill commented on Source/WebCore/Headers.cmake:
```diff
@@ -1832,6 +1832,8 @@ set(WebCore_PRIVATE_FRAMEWORK_HEADERS
     layout/integration/LayoutIntegrationBoxTreeUpdater.h
     layout/integration/LayoutIntegrationUtils.h
     layout/integration/flex/LayoutIntegrationFlexLayout.h
+
```
Was this newline attentional?

yulunwu2 replied:
Yea, this came from the linter:

Source/WebCore/Headers.cmake:1835: error: [list/emptyline] There should be exactly one empty line instead of 0 between "layout/integration/flex/LayoutIntegrationFlexLayout.h" and "layout/integration/grid/LayoutIntegrationGridLayout.h".

sammygill reviewed (approved):
Note that you have a "No new tests (OOPS!)." inside the commit message that you probably want to remove

yulunwu2 committed:
Add initial integration for GridFormattingContext
https://bugs.webkit.org/show_bug.cgi?id=298884
<rdar://160627422>

Reviewed by Sammy Gill.

This PR adds the foundational integration layer for GridFormattingContext.

Specifically, this PR:

1. Adds the integration point for the new Grid Formatting Context in the form of
LayoutIntegrationGridLayout.
2. Introduces layoutUsingGridFormattingContext() and canUseForGridLayout() to gate
usage of the new layout code path.
3. Adds buildTreeForGridContent() to the the BoxTreeUpdater to build a tree of Layout::ElementBox.
4. Caches the top-level GridBox and LayoutState in GridFormattingContext.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/layout/formattingContexts/grid/GridFormattingContext.cpp:
(WebCore::Layout::GridFormattingContext::GridFormattingContext):
* Source/WebCore/layout/formattingContexts/grid/GridFormattingContext.h:
* Source/WebCore/layout/integration/LayoutIntegrationBoxTreeUpdater.cpp:
(WebCore::LayoutIntegration::BoxTreeUpdater::build):
(WebCore::LayoutIntegration::BoxTreeUpdater::buildTreeForGridContent):
* Source/WebCore/layout/integration/LayoutIntegrationBoxTreeUpdater.h:
* Source/WebCore/layout/integration/LayoutIntegrationCoverage.cpp:
(WebCore::LayoutIntegration::canUseForGridLayout):
* Source/WebCore/layout/integration/LayoutIntegrationCoverage.h:
* Source/WebCore/layout/integration/grid/LayoutIntegrationGridLayout.cpp: Copied from Source/WebCore/layout/formattingContexts/grid/GridFormattingContext.cpp.
(WebCore::LayoutIntegration::GridLayout::GridLayout):
(WebCore::LayoutIntegration::m_layoutState):
(WebCore::LayoutIntegration::GridLayout::layout):
* Source/WebCore/layout/integration/grid/LayoutIntegrationGridLayout.h: Copied from Source/WebCore/layout/formattingContexts/grid/GridFormattingContext.h.
(WebCore::LayoutIntegration::GridLayout::gridBox const):
(WebCore::LayoutIntegration::GridLayout::gridBox):
(WebCore::LayoutIntegration::GridLayout::gridBoxRenderer const):
(WebCore::LayoutIntegration::GridLayout::gridBoxRenderer):
(WebCore::LayoutIntegration::GridLayout::layoutState):
(WebCore::LayoutIntegration::GridLayout::layoutState const):
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::layoutBlock):
(WebCore::RenderGrid::layoutUsingGridFormattingContext):
* Source/WebCore/rendering/RenderGrid.h:

Canonical link: https://commits.webkit.org/300088@main
